schema.zmodel

```diff
@@ -0,0 +1,116 @@
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+
+generator client {
+  provider = "prisma-client-js"
+}
+
+// Here's the simplified model
+//   - Permission is changed to enum so that using it doesn't require join models
+
+// Let's use this project to collaborate
+
+enum Permission {
+  ALL
+  CREATE
+  READ
+  UPDATE
+  DELETE
+}
+
+/*
+ * User can be member of many groups, and can operate many posts.
+ */
+model User {
+  id String @id @default(cuid())
+  groups GroupUser[]
+  posts PostUser[]
+  postPermission Permission?
+
+  // a synthesized field to facilitate access policies
+  groupPostPermission Permission? @ignore
+}
+
+/*
+ * Group can have many users, and (its members) can operate many posts.
+ */
+model Group {
+  id String @id() @default(cuid())
+  users GroupUser[]
+  posts PostGroup[]
+  postPermission Permission?
+}
+
+/*
+ * User-Group join model, with permissions indicating what the user can do to the group
+ */
+model GroupUser {
+  id String @id() @default(cuid())
+  group Group @relation(fields: [groupId], references: [id])
+  groupId String
+  user User @relation(fields: [userId], references: [id])
+  userId String
+  permission Permission
+  @@unique([groupId, userId, permission])
+
+  // current user can manage group members if he has ALL permission to the group
+  @@allow('all', group.users?[user == auth() && user.permission == ALL])
+}
+
+/*
+ * Post belongs to many users and many groups
+ */
+model Post {
+  id String @id() @default(cuid())
+  groups PostGroup[]
+  users PostUser[]
+
+  // require login
+  @@deny('all', auth() == null)
+
+  // current user has global ALL permission to posts
+  @@allow('all', auth().postPermission == ALL)
+
+  // current user belongs to a group with global ALL permission to posts
+  @@allow('all', auth().groupPostPermission == ALL)
+
+  // current user is in the user list of this post and has ALL permission
+  @@allow('all', users?[user == auth() && user.permission == ALL])
+
+  // current user belongs to a group of this post that has ALL permission
+  @@allow('all', groups?[group.group.users?[user == auth()] && group.permission == ALL])
+}
```

Comment:
The rules above represent four kinds of permissions:

- Current user can access Post if it has global Post permission
- Current user can access Post if he's a member of a group with global Post permission
- Current user can access Post if he's in the user list of the post with permission
- Current user can access Post if he's a member of the group list of the post with permission

I've only shown "all" permissions here but the others should be the same.

Reply:
The one thing that I don't like, is that I have to relate a `Group` with a `Post`. I just want a `Group` that has a permission to other `models` without having to relate them.

Reply:
In the Discord topic I suggested this `Group.postPermission.permission.all`, to access directly the `model` to make this possible. But I don't if can be implemented. With the `@ignore` being available I think there is another possibility, but I cannot think one right now. What do you think?

Reply:
> 

I see. I actually misunderstood the requirement and thought `Post` can actually be "owned" by `Group`s.

I think we can then drop the 4th item in my original list above. The `Group.postPermission.permission.all` is covered by the #2 in the list, right? It's covered by the extra field added to `auth()`, as shown in the `Post` model:

```
  // current user belongs to a group with global ALL permission to posts
  @@allow('all', auth().groupPostPermission == ALL)
```

The `groupPostPermission` is precomputed to check if the current user belongs to a group with "ALL" post permissions. Is this what you need here?

I'm not very sure if I understand the exact requirement here. I thought it was still indirectly related to the current user, or is it not?

Reply:
Btw, forgot to mention, without the #4 requirement, the `PostGroup` model and its reference from `Post` can be deleted, and the schema got even simpler.

Reply:
My ideia with `Group` it's just a collection of users that can do stuff based on the permission that the `Group` has. And not a `Group` that interact with the models itself. It's just a way to group global permissions

Reply:
> Right, if we remove the `PostGroup` model, `Group` is not related to any model anymore except for `User`, so it's just a collection of user. When checking if the current user can operate on a resource (like `Post`), we precompute if the user belongs to a `Group` with proper permission and use that flag (`groupPostPermission `) to verdict access. Does this make sense?

Yeah, this makes sense. It's a thing that needs to be check before the initialization of the enhanced ZenStack Prisma client like you demonstrate below. So I guess that's it for me right now.